zerver/views/message_fetch.py

```diff
@@ -936,6 +936,64 @@ def parse_anchor_value(anchor_val: Optional[str], use_first_unread_anchor: bool)
         raise JsonableError(_("Invalid anchor"))
 
 
+@has_request_variables
+def get_messages_for_ids(
+    request: HttpRequest,
+    maybe_user_profile: Union[UserProfile, AnonymousUser],
+    message_ids: List[int] = REQ(json_validator=check_list(check_int)),
+) -> HttpResponse:
+    if not maybe_user_profile.is_authenticated:
+        is_web_public_query = True
+        user_profile: Optional[UserProfile] = None
+    else:
+        is_web_public_query = False
+        assert isinstance(maybe_user_profile, UserProfile)
+        user_profile = maybe_user_profile
+        assert user_profile is not None
+
+    if is_web_public_query:
+        realm = get_valid_realm_from_request(request)
+        user_message_flags = {msg_id: ["read"] for msg_id in message_ids}
+        messages = messages_for_ids(
+            message_ids=message_ids,
+            user_message_flags=user_message_flags,
+            search_fields={},
+            apply_markdown=True,
+            client_gravatar=True,
+            allow_edit_history=realm.allow_edit_history,
+        )
+
+        # Stream of all messages requested should be `web public`.
+        for message in messages:
+            if message["type"] != "stream":
+                raise MissingAuthenticationError()
+        stream_ids = set([message["stream_id"] for message in messages])
+        if Stream.objects.filter(id__in=stream_ids, is_web_public=False).count():
```

Comment:
This faster than using `access_stream_by_*` here, so I used the query directly here. One can get an additional info that if a message is in a web public stream or not thought this, which not a hack??